src/tests/eventpipe/common/StreamProxy.cs

```diff
@@ -0,0 +1,114 @@
+        // forward those bytes to the caller. If the caller
+        // would have thrown an exception with its Read, copy
+        // the bytes to the internal MemoryStream and then throw
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null || offset < 0 || count < 0)
+                throw new ArgumentException("Invalid input into Read");
+
+            byte[] localBuffer = ArrayPool<byte>.Shared.Rent(count);
+            var readCount = ProxiedStream.Read(localBuffer, 0, count);
+            if (readCount == 0)
+                return readCount;
+
+            InternalStream.Write(localBuffer, 0, readCount);
+
+            if (buffer.Length - offset < count)
+            {
+                // This is the error that EventPipeEventSource is causing,
+                // so this is when we should throw an exception, just like
+                // System.IO.PipeStream. This will result in the dispose method
+                // being called and the culprit stream data being dumped to disk
+                Logger.logger.Log($"[Error] Attempted to read {count} bytes into a buffer of length {buffer.Length} at offset {offset}");
+
+                // Throw the exception like what would have happened in System.IO.PipeStream
+                throw new ArgumentException($"Attempted to read {count} bytes into a buffer of length {buffer.Length} at offset {offset}");
+            }
+
+            // copy the data into the caller's buffer
+            Array.Copy(localBuffer, 0, buffer, offset, readCount);
+
+            ArrayPool<byte>.Shared.Return(localBuffer, true);
+            return readCount;
+        }
+
+        public override long Seek(long offset, SeekOrigin origin) => ProxiedStream.Seek(offset, origin);
+
+        public override void SetLength(long value) => ProxiedStream.SetLength(value);
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            // This stream is only for "reading" from. No need for this method.
+            throw new System.NotImplementedException();
+        }
+
+        protected bool disposed = false;
+        protected override void Dispose(bool disposing)
+        {
+            if (!disposed)
+            {
+                if (disposing)
+                {
+                    ProxiedStream.Dispose();
+                    InternalStream.Dispose();
+                }
+
+                disposed = true;
+            }
+        }
+
+        public void DumpStreamToDisk()
+        {
+            var helixWorkItemDirectory = System.Environment.GetEnvironmentVariable("HELIX_WORKITEM_UPLOAD_ROOT");
+            if (helixWorkItemDirectory != null && Directory.Exists(helixWorkItemDirectory))
+            {
+                Logger.logger.Log($"\t HELIX_WORKITEM_UPLOAD_ROOT = {helixWorkItemDirectory}");
+                var filePath = Path.Combine(helixWorkItemDirectory, "streamdump.nettrace");
```

Comment:
I have to admit that I'm not familiar with Helix, so I'll ask @hoyosjs to help out here -- Would this be overwritten if there are more than one tests failing and trying to dump the trace? Or would it be fine because each test execution will go into different `helixWorkItemDirectory`? (By "each test execution" I mean each `eventpipe` test harnesses, not test CI legs).

Reply:
Problem is we don't really use Helix. Even if we did, every test assembly will run as a WorkItem. Something like appending a pid might help, but even that is a heuristic (for example, we run in Docker which has very deterministic pid assignments which is problematic). A parameter intake for as a unique identifier would be better.